### PR TITLE
Fabtests/Multinode: Performance

### PR DESCRIPTION
**Pull request by alex-mckinley**
Add options to run only one pattern for the multinode tests

Show performance data for connection management

### CODE REVIEW

File: fabtests/multinode/src/timing.c [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+#include <string.h>
+#include <timing.h>
+#include <core.h>
+#include <limits.h>
+
+extern struct pm_job_info pm_job;
+
+int multi_timer_create(struct multi_timer **timer, int rank)
+{
+	*timer = malloc(sizeof(*timer));
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:41]: Allocating 1 structure at a time is horribly inefficient.

File: fabtests/multinode/src/timing.c [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+void multi_timer_stop(struct multi_timer *timer)
+{
+	struct timespec spec;
+
+	if (timer && timer->start != -1 && timer->end == -1) {
+		clock_gettime(CLOCK_REALTIME, &spec);
+		timer->end = spec.tv_nsec + spec.tv_sec * 1000000000;
+	}
+}
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:68]: We have functions to get timestamps -- ft_gettime_ns() / us() / ms() -- which can be used instead, and call clock_gettime() correctly.  We shouldn't need either of the above start/stop functions.

File: fabtests/include/shared.h [hunk 3]
```diff
@@ -125,6 +125,7 @@ enum {
 	FT_OPT_SRX		= 1 << 21,
 	FT_OPT_STX		= 1 << 22,
 	FT_OPT_SKIP_ADDR_EXCH	= 1 << 23,
+	FT_OPT_PERF	= 1 << 24,
```
Comment (shefty) [inline fabtests/include/shared.h:128]: alignment looks off here, but that could be the web view

File: fabtests/multinode/src/core.c [hunk 4]
```diff
@@ -48,6 +48,7 @@
 #include <core.h>
 #include <pattern.h>
 #include <shared.h>
+#include <timing.h>
```
Comment (shefty) [inline fabtests/multinode/src/core.c:51]: Local include files should be using "x.h" rather than <x.h>

File: fabtests/multinode/src/core.c [hunk 5]
```diff
@@ -87,6 +90,8 @@ static int multi_setup_fabric(int argc, char **argv)
 	hints->ep_attr->type = FI_EP_RDM;
 	hints->mode = FI_CONTEXT;
 	hints->domain_attr->mr_mode = opts.mr_mode;
+	hints->tx_attr->size = pm_job.num_ranks;
+	hints->rx_attr->size = pm_job.num_ranks;
```
Comment (shefty) [inline fabtests/multinode/src/core.c:94]: This change looks unrelated.

File: fabtests/multinode/src/core.c [hunk 6]
```diff
@@ -247,6 +251,11 @@ int multi_msg_send()
 
 		offset = state.sends_posted % opts.window_size;
 		assert(tx_ctx_arr[offset].state == OP_DONE);
+		if (ft_check_opts(FT_OPT_PERF)) {
+			timer_index = state.iter * pm_job.num_ranks +
+				(state.sends_posted % pm_job.num_ranks);
+			multi_timer_start(timers[timer_index]);
```
Comment (shefty) [inline fabtests/multinode/src/core.c:257]: I have no idea what timer_index represents or if this is correct.  I *think* timers would be better defined as some sort of multi-dimensional array, but I can't tell how the array is structured.  For example, is it this (conceptually)?
```
struct timestamp {
    uint64_t start_time;
    uint64_t stop_time;
};

struct timestamp times[RANKS][ITER];
```
in which case, this could be:
```
#define START_INDEX 0
#define STOP_INDEX 1
uint64_t times[RANKS][ITER][2];
```

**Review by shefty — COMMENTED** [review]
Comments on 1st patch only

File: fabtests/multinode/src/core.c (hunk 6 above)
Comment (alex-mckinley) [reply fabtests/multinode/src/core.c:257]: That is basically how it is structured currently. I set it up this way to make it easier to move all the timers through a sockets call. The 2d array is much easier to read, and this is a lot more readable. 

Does the concern about sending the timers even matter?

Comment (shefty) [reply fabtests/multinode/src/core.c:257]: I would wrap the timestamps in some sort of 2D array abstraction.  Something sort of like this:
```
ts_alloc(max_ranks, max_entries)
{
    ts_array = calloc(max_ranks * max_entries, sizeof(*ts_array));
    ts_array->max_ranks = max_ranks;
    ts_array->max_entries = max_entries;
}

ts_entry(ts_array, rank, index)
{
    assert(rank < ts_array->max_rank && index < ts_array->max_entries);
    return ts_array + rank * ts_array->max_entries + index;
}
```
then you could do:
```
ts_entry(ts_array, cur_rank, iter)->start = ft_gettime_ns();
...
ts_entry(ts_array, cur_rank, iter)->stop = ft_gettime_ns();
```
This makes it clearer that we're dealing with an array of timestamps per rank, and removes the questionable use of the modulo operator in finding the entry.

Comment (alex-mckinley) [reply fabtests/multinode/src/core.c:257]: This makes sense, I will reimplement this way

Comment (alex-mckinley) [reply fabtests/multinode/src/core.c:257]: As I'm implementing this, I'm not actually sure that this saves any sort of index magic. In the current implementation the timers are ordered the way they are used, num_ranks per iteration (at maximum, some patterns use fewer). A 2d array implementation does look a tad cleaner, but the modulo just ends up inside of the index directly.

So to start a given timer it would look like: `timer_start(timers[iter][state.sends_posted % pm_job.num_ranks])`, which is slightly cleaner given the lack of multiplication on the first index.

Moving away from all of these structs does make sense in that it saves a lot of malloc calls, but I don't know if it does the readability improvement you're looking for.

Comment (shefty) [reply fabtests/multinode/src/core.c:257]: What I'm looking for is an easy and direct mapping for the array entry that a given rank uses for a specific send.  It's the modulo that's confusing things.  See the ts_entry() function I listed above.  This assumes that we can calculate the total number of timestamps that we need up front in order to do one allocation.

**Review by zachdworkin — CHANGES_REQUESTED** [review]
runmultinode.sh needs a return value so that if it fails it does not fail silently.
Currently if it fails silently the CI will think it passed and approve the stage.

Additionally the log that gets created by runmultinode.sh needs to be cat'ed so that jenkins can view the log and put it into the output of the stage.

**Comment by alex-mckinley** [comment]
@shefty Updated and ready for rereview

File: fabtests/multinode/src/core.c [hunk 7]
```diff
@@ -244,6 +247,10 @@ int multi_msg_send()
 
 		offset = state.sends_posted % opts.window_size;
 		assert(tx_ctx_arr[offset].state == OP_DONE);
+		if (ft_check_opts(FT_OPT_PERF)) {
+			timer_index = (state.iter * pm_job.num_ranks) + state.cur_target;
```
Comment (shefty) [inline fabtests/multinode/src/core.c:251]: The timer index calculation is repeated several times.  Please create a static inline helper function to return this.  E.g.
```
static inline int timer_index(int iter, int dest_rank)
{
    return iter * pm_job.num_ranks + dest_rank;
}
```
assuming I interpreted cur_target correctly.

File: fabtests/multinode/src/core.c [hunk 8]
```diff
@@ -460,18 +479,33 @@ int multinode_run_tests(int argc, char **argv)
 	if (ret)
 		return ret;
 
+	if (ft_check_opts(FT_OPT_PERF))
+		timers = malloc(sizeof(*timers) * opts.iterations * pm_job.num_ranks);
```
Comment (shefty) [inline fabtests/multinode/src/core.c:483]: Use calloc, which is safer.

File: fabtests/multinode/src/core.c [hunk 9]
```diff
@@ -423,15 +438,20 @@ static int multi_run_test()
...
 		if (ret)
 			return ret;
+
 	}
 	return 0;
 }
 
 static void pm_job_free_res()
 {
+	free(timers);
```
Comment (shefty) [inline fabtests/multinode/src/core.c:454]: Here you call free(timers), but below...

File: fabtests/multinode/src/core.c [hunk 10]
```diff
@@ -460,18 +479,33 @@ int multinode_run_tests(int argc, char **argv)
...
+			for (j = 0; j < opts.iterations * pm_job.num_ranks; j++)
+				free(timers[j]);
+		}
 
 		fflush(stdout);
 	}
-
+out:
+	for (j = 0; j < opts.iterations * pm_job.num_ranks; j++)
+		free(timers[j]);
```
Comment (shefty) [inline fabtests/multinode/src/core.c]: ... you call free(timers[j]).  Which is it?  May be better to have timer_alloc/timer_free calls.

File: fabtests/multinode/src/timing.c (hunk 1 above)
Comment (shefty) [reply fabtests/multinode/src/timing.c:42]: Previous comment is still true. Only now I don't understand why there's a malloc here and a different malloc above.

File: fabtests/multinode/src/timing.c [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+extern struct pm_job_info pm_job;
+
+int multi_timer_create(struct multi_timer **timer, int rank)
+{
+	*timer = malloc(sizeof(*timer));
+
+	(*timer)->rank = rank;
+	(*timer)->start = -1;
+	(*timer)->end = -1;
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:46]: It's more efficient to use 0 for an invalid timer.  calloc will initialize the memory to 0, so we don't need a separate init to set it to -1..

File: fabtests/multinode/src/timing.c [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+	double total_timers, total_min, total_max, total_sum_time, total_duration;
+	long *min = malloc(sizeof(*min) * iterations);
+	long *max = malloc(sizeof(*max) * iterations);
+	long *first_start = malloc(sizeof(*first_start) * iterations);
+	long *last_end = malloc(sizeof(*last_end) * iterations);
+	double *sum_time = malloc(sizeof(*sum_time) * iterations);
+	long duration;
+	struct multi_timer **gather_timers;
+	double iter_timer_count;
+	gather_timers = malloc(sizeof(**gather_timers) *
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:76]: nit: add blank line between declarations and other code

File: fabtests/multinode/src/timing.c [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+int multi_timer_analyze(struct multi_timer **timers, int timer_count)
+{
+	int i, j, ret = 0;
+	int iterations = timer_count / pm_job.num_ranks;
+	double total_timers, total_min, total_max, total_sum_time, total_duration;
+	long *min = malloc(sizeof(*min) * iterations);
+	long *max = malloc(sizeof(*max) * iterations);
+	long *first_start = malloc(sizeof(*first_start) * iterations);
+	long *last_end = malloc(sizeof(*last_end) * iterations);
+	double *sum_time = malloc(sizeof(*sum_time) * iterations);
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:72]: All malloc calls would be safer as calloc, and are more natural for calloc anyway.

**Review by shefty — CHANGES_REQUESTED** [review]

**Comment by alex-mckinley** [comment]
updated with feedback

File: contrib/intel/jenkins/Jenkinsfile [hunk 14]
```diff
@@ -294,6 +295,338 @@ pipeline {
                   echo "MPI-tcp-rxm-1 completed."
                 )
               """
+=======
```
Comment (zachdworkin) [inline contrib/intel/jenkins/Jenkinsfile:298]: It doesn't look like you resolved this confict.
You need to keep the current upstream jenkinsfile and just add your multinode-test stage.
Make sure to keep the spacing correct. It is using two spaces as indentation and not tabs.

**Review by zachdworkin — CHANGES_REQUESTED** [review]

**Comment by alex-mckinley** [comment]
Fixed merge conflicts

**Review by zachdworkin — APPROVED** [review]
@shefty, this looks fine from the CI side

**Comment by alex-mckinley** [comment]
Fixed newest set of merge conflicts in the runscripts.py

File: fabtests/multinode/src/timing.c [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+			total_sum_time/total_timers);
+
+err:
+	free(min);
+	free(max);
+	free(sum_time);
+	free(first_start);
+	free(last_end);
+	free(gather_timers);
+	return ret & 1;
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:137]: This is a really weird return code.  Just return 0 on success or -error on failure.  I have no idea what the intent is here.

File: fabtests/multinode/src/timing.c [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+			for (j = 0; j < timer_count; j++)
+				all_timer[i * timer_count + j] = recv_timer[j];
+
+			free(recv_timer);
+		}
+	} else {
+		ret = socket_send(pm_job.sock, timers, 
+						  sizeof(*recv_timer) * timer_count, 0);
+		if (ret < 0)
+			return ret;
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:166]: If check isn't needed as we call ```return ret```  below anyway.

File: fabtests/multinode/src/timing.c [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+	free(max);
+	free(sum_time);
+	free(first_start);
+	free(last_end);
+	free(gather_timers);
+	return ret & 1;
+}
+
+int multi_timer_gather(struct multi_timer *all_timer,
+					   struct multi_timer *timers, int timer_count)
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:141]: alignment

File: fabtests/multinode/src/timing.c (hunk 15 above)
Comment (alex-mckinley) [reply fabtests/multinode/src/timing.c:137]: So I do think I have an error here, but the reason I did this was because this ret is basically set by a sockets call and that value gets passed back to ft_exit_code, which does not like a positive value and incorrectly concludes that the tests failed. I was trying to mitigate that, but this is not the right way

**Comment by alex-mckinley** [comment]
Fixed latest set of requested changes

File: fabtests/multinode/src/timing.c [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+	long *max = calloc(iterations, sizeof(*max));
+	long *first_start = calloc(iterations, sizeof(*first_start));
+	long *last_end = calloc(iterations, sizeof(*last_end));
+	double *sum_time = calloc(iterations, sizeof(*sum_time));
+	long duration;
+	struct multi_timer *gather_timers;
+	double iter_timer_count = 0;
+
+	gather_timers = calloc(pm_job.num_ranks * pm_job.num_ranks, 
+							sizeof(*gather_timers));
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:78]: The indentation here is painful to my eyeballs.

File: fabtests/multinode/src/timing.c [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+	
+	if (!(min && max && first_start && last_end && sum_time && gather_timers))
+		goto err;
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:81]: This should set ret -FI_ENOMEM before jumping to err.

File: fabtests/multinode/src/timing.c [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+		pm_barrier();
+	}
+
+	if (pm_job.my_rank == 0)
+		printf("%-10s %16.3lf %16.3lf %16.3lf %16.3lf\n", "Average",
+			total_min/iterations, total_max/iterations,
+			total_duration/iterations,
+			total_sum_time/total_timers);
+
+err:
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:130]: This is a common exit path, so I'd change the label from err to out.  Also if you set ret = 0 immediately before the label, that will clearly indicate that the function was successful and allow you to remove the if-else statement below.

File: fabtests/multinode/src/timing.c [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+			for (j = 0; j < pm_job.num_ranks * pm_job.num_ranks; j++) {
+				if (gather_timers[j].start == 0 || gather_timers[j].end == 0)
+					continue;
+
+				iter_timer_count++;
+				duration = gather_timers[j].end - gather_timers[j].start;
+				sum_time[i] += duration;
+
+				if (gather_timers[j].start < first_start[i] 
+					|| first_start[i] == 0)
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:101]: Please re-format to move the '||' to the line above and align first_start directly under gather_timers.
```
if (gather_timers... ||
_   first_start ...)
```
ignore the '_' here -- that's a hack to show the alignment without markdown converting my spaces to a tab.

File: fabtests/multinode/src/timing.c [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+			"Iteration", "Min Send (ns)", "Max Send (ns)", 
+			"Pattern Time(ns)", "Average Send(ns)");
+
+	for (i = 0; i < iterations; i++) {
+		ret = multi_timer_iter_gather(gather_timers, timers, i);
+		if (ret < 0)
+			goto err;
+		if (pm_job.my_rank == 0) {
+			for (j = 0; j < pm_job.num_ranks * pm_job.num_ranks; j++) {
+				if (gather_timers[j].start == 0 || gather_timers[j].end == 0)
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:93]: Here and above, please check line lengths and limit to about 80 characters.  A couple characters over is fine if it helps with readability.

File: fabtests/multinode/src/timing.c [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+				sum_time[i] += duration;
+
+				if (gather_timers[j].start < first_start[i] 
+					|| first_start[i] == 0)
+					first_start[i] = gather_timers[j].start;
+				if (gather_timers[j].end > last_end[i])
+					last_end[i] = gather_timers[j].end;
+				if (duration > max[i])
+					max[i] = duration;
+				if (duration < min[i] || min [i] == 0)
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:107]: remove space between min and [i]

File: fabtests/multinode/src/timing.c [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+void multi_timer_stop(struct multi_timer *timer)
+{
+	if (timer && timer->start != 0 && timer->end == 0)
+		timer->end = ft_gettime_ns();
+}
+
+static inline void print_timer(struct multi_timer timer, char* info)
+{
+	printf("rank: %i, start: %ld, end: %ld, %s\n", 
+			timer.rank, timer.start, timer.end, info);
```
Comment (shefty) [inline fabtests/multinode/src/timing.c:60]: align timer.rank under "rank

**Review by shefty — COMMENTED** [review]
Assume I'm absolutely anal about indentation, line length, and text alignment.

The problem is that ends up drawing your eyes away from focusing on the code and instead to the alignment, which slows down the actual reading of the code.

**Comment by alex-mckinley** [comment]
Formatting updated

File: fabtests/multinode/src/harness.c [hunk 25]
```diff
@@ -57,7 +57,23 @@ static int parse_caps(char* caps)
...
+	} else if (strcmp(pattern, "ring") == 0) {
+		return 1;
+	} else if (strcmp(pattern, "gather") == 0) {
+		return 2;
+	} else if (strcmp(pattern, "broadcast") == 0) {
+		return 3;
+	} else {
+		printf("Warn: Invalid pattern, defaulting to full_mesh\n");
+		return 0;
+	}
```
Comment (shefty) [inline fabtests/multinode/src/harness.c:73]: As a follow up change, please convert the pattern to an enum, so we don't need a magic decoder ring to determine that '2' is gather.  E.g.
```
enum mnode_pattern {
    PATTERN_MESH,
    PATTERN_RING,
    ...
};

enum mnode_patter pattern;
```

**Review by shefty — APPROVED** [review]
